Respawn delay

contradict opened the pull request:
Sometimes, especially for hardware drivers, it helps to pause before restarting a failed node.

This branch adds a configurable delay to the roslaunch respawn feature. If the delay is left unspecified, it defaults to 0, giving the same behavior as before. The value is configured in the node tag:
<node
     name="myname"
     pkg="flaky_hardware"
     type="unreliable_node"
    respawn="True"
    respawn_delay="5.0"
/>
respawn_delay has units seconds.

contradict committed:
Add respawn_delay parsing and storage

Modify parser and all relevant classes to keep track of optional
respawn_delay. No behavioral changes.

contradict committed:
Track time of node shutdown

Record the first time which is_alive() returns False. Using this
as time_of_death assumes is_alive is being polled, fortunately this
is exactly what ProcessMonitor does.
time_of_death is initialized to None and reset to None on node startup

contradict committed:
respawn after a configurable delay

Keep a list of nodes which have exited and whos respawn_delay has not yet
passed. Check the list each time through the process monitoring loop, restarting
nodes as their delay expires.

ros-pull-request-builder commented:
Can one of the admins verify this patch?

dirk-thomas commented:
@ros-pull-request-builder

ros-pull-request-builder commented:
Test failed.
Refer to this link for build results: http://jenkins.ros.org/job/_pull_request-indigo-ros_comm/28/

contradict committed:
Move time_of_death handling up to Process class

I had this in LocalProcess, base on the testing architecture it looks
like it should go here.

contradict committed:
Add respawn_delay to unit tests

contradict committed:
Correct mock behavior

Needs to track time_of_death just like the real Process object

contradict committed:
Fix respawn test

Was never actually asked to respawn, and no check was made to see if it did.

contradict committed:
Add test for respawn delay

ros-pull-request-builder commented:
Test passed.
Refer to this link for build results: http://jenkins.ros.org/job/_pull_request-indigo-ros_comm/29/

dirk-thomas commented:
Please update the PR so that added/changed lines comply with PEP 8 and avoid any non-necessary whitespace changes in order to keep the diff minimal.

contradict committed:
Split long lines

contradict committed:
Remove space inside parens

as per PEP-8

contradict committed:
Undo unnecessary whitespace change

contradict commented:
Can you be more specific? I took a crack at this and found that most of what I could do was breaking long lines. Most of the existing code uses very long lines, so I'm not sure this is a good change. I found one line where I unnecessarily changed the indentation and a couple where I used space inside parens in a non-pep 8 way. Is there some aspect of the formatting you would particularly like me to fix?

ros-pull-request-builder commented:
Test failed.
Refer to this link for build results: http://jenkins.ros.org/job/_pull_request-indigo-ros_comm/30/

contradict committed:
Fix assertion broken during split of long lines

ros-pull-request-builder commented:
Test passed.
Refer to this link for build results: http://jenkins.ros.org/job/_pull_request-indigo-ros_comm/31/

dirk-thomas commented on tools/roslaunch/src/roslaunch/pmon.py:
```diff
@@ -538,15 +558,15 @@ def _run(self):
-                        logger.debug("Process[%s] has died, respawn=%s, required=%s, exit_code=%s",p.name, p.respawn, p.required, p.exit_code)
+                        logger.debug("Process[%s] has died, respawn=%s, required=%s, exit_code=%s",
+                                p.name,
+                                "True(%f)"%p.respawn_delay if p.respawn else p.respawn,
```
Please add spaces around `%`.

dirk-thomas commented on tools/roslaunch/src/roslaunch/pmon.py:
```diff
@@ -582,18 +604,23 @@ def _run(self):
-                    # stop process, don't accumulate errors
-                    r.stop([])
-                    r.start()
+                    if r.should_respawn()<=0.0:
```
Please add spaces around `<=`.

dirk-thomas commented on tools/roslaunch/src/roslaunch/pmon.py:
```diff
@@ -582,18 +604,23 @@ def _run(self):
-                    r.stop([])
-                    r.start()
+                    if r.should_respawn()<=0.0:
+                        printlog("[%s] restarting process"%r.name)
```
Spaces around `%`.

dirk-thomas commented on tools/roslaunch/test/unit/test_roslaunch_pmon.py:
```diff
@@ -427,6 +462,12 @@ def f():
+        self.failIf(p3.spawn_count < 2, "process did not respawn")
+
+        self.failIf(p4.respawn_interval < p4.respawn_delay,
+                "Respawn delay not respected: %s %s"%(p4.respawn_interval,
```
Spaces around `%`.

dirk-thomas commented on tools/roslaunch/src/roslaunch/core.py:
```diff
@@ -514,6 +518,7 @@ def xmlattrs(self):
             ('output', self.output),
             ('cwd', cwd_str), 
             ('respawn', self.respawn), #not valid on <test>
+            ('respawn_delay', self.respawn_delay), #not valid on <test>
```
A trailing comment should have two spaces before the `#` and one after.

contradict committed:
Spaces around operators

Reformat to include whitespace around operators and comment marks
in new code.

ros-pull-request-builder commented:
Test failed.
Refer to this link for build results: http://jenkins.ros.org/job/_pull_request-indigo-ros_comm/33/

dirk-thomas commented:
Please rebase the PR to the tip of the branch.

ros-pull-request-builder commented:
Test failed.
Refer to this link for build results: http://jenkins.ros.org/job/_pull_request-indigo-ros_comm/34/

contradict commented:
I have so far been unable to duplicate this build failure on my system. I tried running the tests using the jenkins_tools package, setting up my own containerized environment, and running the test in my development environment. In all cases, the test which fails here runs successfully. The raw log of the test run above indeed looks different than a successful run, it emits the same error message as a successful run but much more frequently. What can I do to more accurately replicate the test environment?

dirk-thomas commented:
You can only run the tests locally (`make run_tests`). There is currently no good way to resemble the exact environment from the build farm.

I will check if I see anything obvious but might not have enough time to get to a useful result.

dirk-thomas commented:
@ros-pull-request-builder

ros-pull-request-builder commented:
Test passed.
Refer to this link for build results: http://jenkins.ros.org/job/_pull_request-indigo-ros_comm/61/

wjwwood commented:
+1

dirk-thomas commented:
Squashed and merged in c171283e548cfdbe5bdafeb5559462beff529a89

dirk-thomas commented:
@contradict Can you please update the documentation in the wiki to mention the new attribute introduced in _Indigo_ (and post the link to the wiki changes here)?

contradict commented:
Update added here, I hope this is the correct place: http://wiki.ros.org/roslaunch/XML/node

dirk-thomas commented:
Thanks, I have updated the page to mention that the attribute has been introduced in Indigo: http://wiki.ros.org/action/diff/roslaunch/XML/node?action=diff&rev1=20&rev2=21
